core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -725,6 +728,21 @@ private static SortOrder freshSortOrder(int orderId, Schema schema, SortOrder so
     return builder.build();
   }
 
+  private Map<String, SnapshotRef> validateRefs(Map<String, SnapshotRef> inputRefs) {
+    for (Map.Entry<String, SnapshotRef> refEntry : inputRefs.entrySet()) {
+      String refName = refEntry.getKey();
+      SnapshotRef ref = refEntry.getValue();
+      if (ref.snapshotId() == -1) {
+        Preconditions.checkArgument(refName.equals(SnapshotRef.MAIN_BRANCH),
+            "Snapshot ref %s refers to snapshot id -1. Only main can refer to -1", refName);
```

Comment:
I think this should be `if (ref.snapshotId() == -1 && refName.equals(MAIN_BRANCH))` then ignore it. Otherwise, check the other precondition. No need to have a special error message for -1 that states main can be -1.

Also, for @jackye1995, do we want to just allow main to be dropped? If you remove main then current-snapshot-id is -1 but we don't actually have -1 in this map?

Reply:
Removed the need for a special error message.

Will let @jackye1995 comment, but would we be putting any implications (i.e. potentially unwanted behaviors) on retention if we allowed main to be dropped? If we allow main to be dropped this means that we must have a fallback mechanism for any other reference types (which would be the retention already defined at the table level). 

This seems like it should be okay, but just trying to think through if there are any cases where allowing someone to drop main leads to any undesired behavior in retention.

Reply:
> If you remove main then current-snapshot-id is -1 but we don't actually have -1 in this map?

I remember we were thinking about deprecating `current-snapshot-id` after the branch change, because `current-snapshot-id` should now be retrieved by getting the snapshot ID of the main branch, and that could be -1 if not exist. If we go with that route, we allow a `{"main": -1}` entry in the map and we can deprecate `current-snapshot-id`. If we keep `current-snapshot-id`, then that information of `{"main": -1}` is duplicate and we don't need to keep that. Either way is fine to me, do you think any one of them has any particular advantage?

Reply:
keeping `current-snapshot-id` has the advantage that an old reader can continue to read the new metadata and get current snapshot id. In that sense, we will likely not be able to retire the field `current-snapshot-id`. So we should probably go with the logic of not having `main` in the map, and if people do `table.ref("main")` it should return `table.currentSnapshot().snapshotId()` or -1

Reply:
I agree. We should not explicitly track -1, but when the current snapshot is requested we can return null if there is no `main`. And we should keep `current-snapshot-id` up to date for older readers.